Deriv tests 2023-03

markshowalter opened the pull request:
This addresses a number of backplane and surface issues, plus scattered other improvements. To run the gold tests, you will need to
  export OOPS_GOLD_MASTER_PATH="$OOPS_RESOURCES/gold_master-deriv-tests-2023-03/"
because that is the temporary location of the revised backplane arrays.

Key changes:
- Numerous new tests of derivatives and a few bug fixes, originally motivated by the weird "JUPITER finest resolution" array. That array is now different, although still weird. They key difference is that I understand what's going on now. We can discuss when Joe is in Mountain View because it's way easier to explain with pictures on a whiteboard.
- You should not see any more "failed to converge" warnings, and many algorithms now converge in way fewer iterations than before. (I am _extremely_ proud of this!)
- There was a "quirky" definition of the altitude ("z") coordinate in some Limb and Ellipsoid surfaces. z is now rigorously defined to be measured from the local surface normal.
- A new inheritance hierarchy among Surface subclasses enabled me to delete a lot of duplicated code.
- With Joe's input, there are a few new, useful options for gold_master.py.
    --derivs compares a bunch of numerical derivatives against the calculated derivatives. This can be slow, so it's not on by default.
    --summary (renamed from --info) summarizes the properties of each backplane array in the log.
    --fullpaths includes the full path to each individual file in the log.
    --save-sampled or --ss saves subsampled copies of the gold master arrays in your local backplane directory.

Other, more minor changes:
- I've started trying to standardize error messages to be more informative. Those are most of the changes inside polymath, and I have also upgraded the error messages in the surface classes. Others can be done over time.
- In the surface module, convergence messages now have units. Useful! This can be added to other modules as we go.
- I added backplanes "mu" and "mu0" because I got tired of typing "emission_angle().cos()" and "incidence_angle().cos()".
- I moved the lambert_law, minnaert_law, and lommel_seeliger_law backplane methods from spheroid.py to lighting.py, because that's where the belong (and because spheroid.py is enormous).
- The old exercise_backplane files have been moved to ideas/backplanes and some of that has been deleted from hosts.
- I found a few gold_master tests that were really just duplicates under different names, so a few of them have been renamed or deleted.
- As usual, scattered bugfixes.

jnspitale reviewed (approved):
Looks good to me.  No unit test failures.  I'm still getting this warning, which is not among the things you said you fixed.  Do we need to make this an issue?

runTest (hosts.solar.unittester.Test_Solar) ... <frozen importlib._bootstrap>:219: RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 80 from C header, got 96 from PyObject

rfrenchseti reviewed (changes requested):
- Voyager ISS no longer works because it imports:

```
from oops.backplane.exercise_backplanes import exercise_backplanes
from oops.backplane.unittester_support  import Backplane_Settings
```

- I'm not fond of the new logfile options that give summaries and filenames as separate lines. That breaks the nice machine-readable log file format, since the added lines don't have log prefixes on them. Personally I would log them like any other lines, using the full prefix information. Then if we ever want to feed them through an automatic log analyzer we will be able to.
- I should've asked this before, but I'm confused by the whole `gold_master` thing. Are these tests run as part of the normal `oops/unittester`?
- The new `--derivs` option to `gold_master` doesn't run noticeably slower for me. Maybe 1 second. I'm not sure why this isn't on by default.
- Just for reference, code coverage on oops has dropped from 76% to 69%.
- Fixing above Voyager issues, my tests pass.
-

markshowalter commented on hosts/cassini/iss.py:
```diff
@@ -417,52 +418,52 @@ def reset():
-if __name__ == '__main__':
-    backplane_unittester_args()
-    unittest.main(verbosity=2)
+#
```
I talked to Joe about this. Right now, there are no existing unit tests for numerous host modules, including cassini.iss. There is an open task to create them, which Joe has assigned to himself. I didn't want to delete whatever skeletal unit tests are there, because that info will help Joe to create the new tests based on gold_master.py. The case of Voyager ISS is a good example, where whatever unit test was ever there has failed because of the missing imports. I have now commented that code out as well, and will check it in shortly.

markshowalter commented on hosts/solar/kurucz.py:
```diff
@@ -28,7 +28,7 @@
 RSUN = 695700.
 FACTOR = 4 * np.pi * (RSUN/solar.AU)**2
 
-if 'FLUX_DENSITY' not in globals(): # pragma: no cover
```
The file is only read if kurucz.py is imported, and only the first time it is imported. The statement

__all__ = []

at the top of hosts/solar/__int__.py ensures that none of the models are imported by default. This module only gets imported it if you specify model="kurucz" as your input to one of the solar methods, which triggers the import of hosts.solar.kurucz. Follow? I agree this is a bit tricky, and have added a longer and clearer explanation before the call to importlib.import_module() in flux_density():

    # Each reference to a named model triggers the import of its associated
    # Python file hosts/solar/<name>.py, referenced as "hosts.solar.<name>"
    # here. Below, the variable "module" is assigned to this module.
    #
    # Each solar module has three required attributes: UNITS, XUNITS, and
    # FLUX_DENSITY. Note that the module's FLUX_DENSITY attribute is sometimes
    # defined by reading a local (and potentially large) data file. If so, that
    # Python file is written to ensure that the data file is only read once, and
    # that subsequent references to module.FLUX_DENSITY will refer to copy of
    # the array that is cached internally. This ensures that each data file is
    # only read if it is needed, and never more than once.

markshowalter commented:
I think I have addressed Rob's comments.

(1) I modified gold_master.py logging as follows:
- I removed the --summary option; that info is now always included. I think this makes sense.
- The --fullpath lines now have the standard log prefix. I am trying to strike a balance between machine-readability and human-readability of the logs (with some emphasis on the latter) but I think they remain human-friendly just by inserting a blank line after each set of rows listing file paths, so that's what I have done. Any impact of a log analyzer would be minimal.

(2) The file oops/backplane/unittester.py runs the standard gm test as a unit test. This executes by default in oops/unittester.py.

(3) I am somewhat surprised by the minimal impact of "--derivs", because it is evaluating literally almost five times as many backplanes. When undersamping = 1, which I use a lot, the difference is 900s vs. 300s. My solution is to have derivative testing turned off by default if undersampling == 1, on by default otherwise. This is explained in the help info. That way I don't have to think about it too much. It is on by default for unit tests now.

(4) I can't think of a good reason for the coverage drop. I deleted a lot of nearly-duplicated lines of surface subclass code, so perhaps one factor is the smaller denominator. Still, I'm surprised, but maybe this is a question for another day.

(5) Voyager ISS tests are now commented out like the others.

rfrenchseti reviewed (approved):
One tiny comment: when doing multi-line log entries I find it nice to slightly indent the subsequent lines so they end up more visually grouped.
